.ci/conda-envs/README.md

```diff
@@ -0,0 +1,11 @@
+# conda-envs
+
+This directory contains files used to create `conda` environments for development
+and testing of LightGBM.
+
+The `.txt` files here are intended to be used with `conda create --file`.
```

Comment:
I'm not really convinced why a `.txt` file would be more appropriate than a `.yml` file here 🤔 I've been happily using `.yml` files for environment creation for years although I've been using `micromamba` instead of `conda`.

While I think that the former would be more fitting for use in CI in general, does anything prevent us from using `conda env create -f <file>.yml`? I'm almost certain that this works.

Reply:
I've also used `micromamba install` on the base env which allows you to specify the python version and a yaml file for the dependencies ([example](https://github.com/Nixtla/neuralforecast/blob/0a772c5a43fc4275ab5c6ed7670ff641d41be83f/.circleci/config.yml#L11))

Reply:
> *does anything prevent us from using conda env create -f <file>.yml?*

I described this in "Why are these .txt files and not conda .yaml files?" in the PR description. I'll add more details here.

I wanted to be able to pass in a Python constraint from the command line, to avoid these alternatives I could think of for constraining Python version:

* using templated statements like `python=${PYTHON_VERSION}` in an `env.yml` file and needing to involve some templater like `envsubst` ([docs](https://man7.org/linux/man-pages/man1/envsubst.1.html)) to use them
* generating individual constraint for each Python version LightGBM tests against, each containing a dependency like `python=3.10.*`
* continuing to have these lists of dependency requirements hard-coded in CI scripts

As of the latest `conda` ([v24.3.0](https://github.com/conda/conda/releases/tag/24.3.0)), `conda env create` does not support mixing files and command-line constraints.

Given a file `env.yaml` with these contents:

```yaml
channels:
  - nodefaults
  - conda-forge
dependencies:
  - scipy>=1.0
```

This

```shell
conda create \
   --name delete-me \
   --file ./env.yaml
```

Yields this

```text
CondaValueError: could not parse '- nodefaults' in: ./env.yaml
```

`conda env create` can read files like that... but you can't pass additional constraints from the command line.

```shell
conda env create \
   --name delete-me \
   --file ./env.yaml \
   python=3.10
```

```text
SpecNotFound: Invalid name 'python=3.10', try the format: user/package
```

I looked some more into this tonight and found this relevant issue with discussion about changes to these APIs in `conda`:

* https://github.com/conda/conda/issues/11465 = request to supporting templating `env.yaml` files (was closed with no changes to `conda`)
* https://github.com/conda/conda/issues/11341 = meta-issue covering various changes to environment creation from YAML files
* https://github.com/conda/conda/issues/9506 = feature request from 2019 that was closed with the suggestion to do exactly what I'm proposing in this PR ... `conda create` with a `.txt` file

> *`micromamba install` ... allows you to specify the python version and a yaml file for the dependencies*

Oh cool! I didn't realize `micromamba` allowed for that. That behavior looks like exactly what I'm trying to get here.

*however* ... I don't support switching to `micromamba` here in LightGBM's CI. It's described as a "reimplementation" of `conda` / `mamba` in its docs ([docs link](https://github.com/mamba-org/mamba?tab=readme-ov-file#micromamba)). Part of the motivation of moving to environment files in this PR is to help make the development of LightGBM easier. I'd really like to avoid having that experience start with "install this specific other `conda` alternative". I also really would not be excited about the prospect of rebuilding the images from https://github.com/guolinke/lightgbm-ci-docker to include `micromamba`.

Especially just for the benefit of "use .yml files because .txt files seem weird".

@jmoralez @borchero can you think of any other *functional* reasons that what I've proposed in this PR (using `conda create` + .txt files + `python=${PYTHON_VERSION}` from the command line) is problematic?

Reply:
I just don't think txt files are widely adopted in the conda commands. For example, suppose someone already has an environment with LightGBM in it and just wants to add the dependencies to develop, they could just use `conda/mamba env update -f env.yaml`. `conda env update` also allows passing multiple files, so if we end up having more than one file they could be passed in that single command.

The two step creation (python version first and dependencies second) isn't really problematic (I think that's what most people do) and I don't think it slows down the second resolution.

Reply:
In my experience, it absolutely does speed up the total time-to-environment-ready to do everything in one step instead of creating an environment and then updating it: #5743.

Reply:
Those times are for the full the full job execution, right? I'd prefer to increase maybe a couple of minutes in the env creation (assuming those numbers still hold with mamba) than introducing a not so familiar and not so flexible conda workflow for contributors.

Also it seems like we're already using micromamba in some jobs ([logs](https://github.com/microsoft/LightGBM/actions/runs/8683164037/job/23808736323?pr=6375#step:3:102)).

Reply:
> Also it seems like we're already using micromamba in some jobs ([logs](https://github.com/microsoft/LightGBM/actions/runs/8683164037/job/23808736323?pr=6375#step:3:102)).

Oh interesting, so maybe it gets installed alongside `mamba` and `conda` if you use miniforge?

In all the times I've seen that ASCII art you linked to, I never once understood it to mean "micromamba". I always thought it was "pmamba" and never stopped to think about what that was 🙃 

```text
          __  ______ ___  ____ _____ ___  / /_  ____ _
         / / / / __ `__ \/ __ `/ __ `__ \/ __ \/ __ `/
        / /_/ / / / / / / /_/ / / / / / / /_/ / /_/ /
       / .___/_/ /_/ /_/\__,_/_/ /_/ /_/_.___/\__,_/
      /_/
```

> than introducing a not so familiar and not so flexible conda workflow for contributors.

I don't agree with this characterization. BUT... I care more about getting these floors on dependencies into the environment solves than I do about the particular mechanism. I'll concede, let's do YAML files.

Would you support still preserving the "create environment in all one shot" approach by having one YAML file per Python version, like this?

```text
conda-envs/
    ci-core-py3.7.yml
    ci-core-py3.8.yml
    ci-core-py3.9.yml
    ci-core-py3.10.yml
```

And then CI scripts would do this.

```shell
conda create --name test-env --file ./.ci/conda-envs/ci-core-py${PYTHON_VERSION}.yml
``

Reply:
Haha yeah I think that's a greek mu. For the single command I prefer the current approach, so I just approved.

Reply:
Thanks!

With the constraints "create the environment with all dependencies in a single command" and "keep using `conda` / `mamba`", there are 2 approaches we've identified:

1. use `conda create --file .ci/ci-core.txt python=${PYTHON_VERSION}`
2. create one YAML file per combination and use something like `conda env create --file ./.ci/conda-envs/ci-core-py${PYTHON_VERSION}.yml`

If I'm reading this correctly, @jmoralez prefers Option 1 and you prefer Option 2.

I also slightly prefer Option 1, at least right now where all not-Python-3.7 environments can share the same dependency constraints.

If we were to have more Python-version-specific lists of dependencies, I think I'd prefer Option 2 more.

@borchero is it alright with you if we move forward with the Option 1 approach right now and see how it goes? Since this is just affecting CI and commands that we document in a README, I think it should be low-risk to be wrong and end up changing this pattern in the future.

Reply:
Yep, fine for me, please move ahead with option 1 to improve the CI setup :)

On a slightly related note, this PR prompted me to think about the use of lock files... this would ensure entirely reproducible environments (i.e. fewer CI failures) and would essentially fully alleviate the need for solving in CI jobs. Potentially something for a future PR... :smile:

Reply:
Nice, I didn't know about the `[RFC]`-type issues yet, I'll think about it a bit more (taking into account your comment, thanks for the elaboration ;) and might come back to this then :smile: